[WIP] Update My Health link to perform account validation before redirecting

narin opened the pull request:
## Description
Currently, when authenticated users click the "My Health" link in the main nav, they are automatically redirected to MHV, even if their account is not in the appropriate state to access it. This newly created page will act as an intermediary that will check their account status, and display any related messaging if they need to take action before being able to access MHV.

Upon clicking the My Health link, users will be redirected to `/my-health-account-validation`, which will fetch their MHV account info on load. If all is well, they will be redirected to MHV as normal.  If not, they will be redirected to one of the pages listed in the table below, depending on the scenario.

This body of work currently has an `isProduction()` feature flag set, so that we can test on dev and staging.

### Scenarios and URLs
| **Scenario**                            | Technical Condition                                          | **URL**                                                      | **Mockup**                                                   |
| --------------------------------------- | ------------------------------------------------------------ | ------------------------------------------------------------ | ------------------------------------------------------------ |
| Verify your identity                    | `!profile.verified` or `accountState === 'needs_identity_verification'` | /my-health-account-validation/verify                         | [Mockup Link](<https://adhoc.invisionapp.com/share/9XRBC0J4TEH#/screens/356934699>) |
| Create a MHV account                    | `!mhvAccount.accountLevel`                                   | /my-health-account-validation/create-account                 | [Mockup Link](<https://adhoc.invisionapp.com/share/9XRBC0J4TEH#/screens/356934701>) |
| Upgrade MHV account                     | `mhvAccount.accountLevel === 'Basic'`                       | /my-health-account-validation/upgrade-account                | [Mockup Link](<https://adhoc.invisionapp.com/share/9XRBC0J4TEH#/screens/358801575>) |
| Accept Terms & Conditions (in progress) | `accountState === 'needs_terms_acceptance'`                  | Update to design being captured in [#17555](https://github.com/department-of-veterans-affairs/vets.gov-team/issues/17555) | [Mockup Link](<https://adhoc.invisionapp.com/share/9XRBC0J4TEH#/screens/355597597>) |
| Verify failed                           | `accountState === 'needs_identity_verification'`             | /my-health-account-validation/error/needs-ssn-resolution     | [Mockup Link](<https://adhoc.invisionapp.com/share/9XRBC0J4TEH#/screens/356934703>) |
| Create account failed                   | `accountState === 'register_failed'`                         | /my-health-account-validation/error/register-failed          | [Mockup Link](<https://adhoc.invisionapp.com/share/9XRBC0J4TEH#/screens/356934704>) |
| Upgrade account failed                  | `accountState === 'upgrade_failed'`                          | /my-health-account-validation/error/upgrade-failed           | [Mockup Link](<https://adhoc.invisionapp.com/share/9XRBC0J4TEH#/screens/358800174>) |
| MHV account inactive                    | `accountState === 'has_deactivated_mhv_ids'`                 | /my-health-account-validation/error/has-deactivated-mhv-ids  | [Mockup Link](<https://adhoc.invisionapp.com/share/9XRBC0J4TEH#/screens/356934706>) |
| Multiple MHV accounts                   | `accountState === 'has_multiple_active_mhv_ids'`             | /my-health-account-validation/error/has-multiple-active-mhv-ids | [Mockup Link](<https://adhoc.invisionapp.com/share/9XRBC0J4TEH#/screens/356934709>) |
| MVI/MHV is down                         | `profile.status === 'SERVER_ERROR'` or `mhvAccount.errors`     | /my-health-account-validation/error                          | [Mockup Link](<https://adhoc.invisionapp.com/share/9XRBC0J4TEH#/screens/356934707>) |
| No VA patient record found              | `accountState === 'needs_va_patient'`                        | /my-health-account-validation/error/needs-va-patient         | [Mockup Link](<https://adhoc.invisionapp.com/share/9XRBC0J4TEH#/screens/356934710>) |



## Testing done
Tested locally

## Acceptance criteria
- [ ] Clicking the "My Health" link redirects user to `/my-health-account-validation`
- [ ] Copy is in line with mockups
- [ ] Users who have the proper account status will be redirected to MHV

## Definition of done
- [ ] Events are logged appropriately
- [ ] Documentation has been updated, if applicable
- [ ] A link has been provided to the originating GitHub issue (or connected to it via ZenHub)
- [ ] No sensitive information (i.e. PII/credentials/internal URLs/etc.) is captured in logging, hardcoded, or specs

Commit:
Intercept My Health click and fetch mhv with user profile

Commit:
Added My Health modal and content

Commit:
Verify my health entry and manifest.  Content update.

Commit:
Content update

Commit:
Update routing

Commit:
Routing updates

Commit:
Use feature flag. Add additional scenario routes

Commit:
Remove old additions from original modal design

Commit:
Remove old additions from original modal design

Commit:
Remove old additions from original modal design

Commit:
Remove old additions from original modal design

Commit:
Remove old additions from original modal design

Commit:
Update feature flag. Rename VerificationFailed content

Commit:
Add staging/prod redirect to mhv

narin commented:
**Note:** This work is feature flagged using `!isProduction()`.  Would like to get this deployed to dev/staging soon so that @tmwolfe1 can begin QA of copy.

Will be using betamocks to mock the different account states in upcoming API PR.

ncksllvn commented on src/applications/validate-mhv-account/routes.jsx:
```diff
@@ -0,0 +1,18 @@
+    { path: 'upgrade-account', component: UpgradeMHVAccount },
+    { path: 'error', component: ErrorMessage },
+    { path: 'error/:errorCode', component: ErrorMessage },
+  ],
```
Should there be a parent component here, ensuring that the user is logged-in or that their account meets certain criteria? Currently, all of these pages are publicly-accessible except for the index route.

ncksllvn commented on src/applications/validate-mhv-account/containers/ValidateMHVAccount.jsx:
```diff
@@ -0,0 +1,111 @@
+
+    // LOA Checks
+    if (!profile.verified) {
+      window.location = 'verify';
```
I think we should possibly be using [`window.location.replace`](https://developer.mozilla.org/en-US/docs/Web/API/Location/replace) here, so that the back button doesn't act funny.

ncksllvn reviewed (approved):
A few suggestions but looks like great work! 

Since this uses client-side routing, I think you'll have to register your route in the [devops config](https://github.com/department-of-veterans-affairs/devops/blob/master/ansible/deployment/config/revproxy-vagov/vars/react_routes.yml#L65) so that your application's landing page is loaded properly.

narin commented:
Merging in order to expedite QA of copy.  Will address @ncksllvn in a new PR.

U-DON commented on src/applications/validate-mhv-account/components/errors/CreateAccountFailed.jsx:
```diff
@@ -0,0 +1,98 @@
+              </p>
+              <p>
+                Tell the representative that you tried to sign in to use health
+                tools on VA.gov, but received an error messaging telling you
```
Is this supposed to be "message" rather than "messaging"?

U-DON commented on src/applications/validate-mhv-account/components/errors/MultipleMHVIds.jsx:
```diff
@@ -0,0 +1,89 @@
+              </p>
+              <a href="https://www.myhealth.va.gov/mhv-portal-web/contact-us">
+                Go to the My HealtheVet contact form
+              </a>{' '}
```
What's with the trailing space here?

U-DON commented on src/applications/validate-mhv-account/containers/CreateMHVAccount.jsx:
```diff
@@ -0,0 +1,59 @@
+
+import {
+  createAndUpgradeMHVAccount,
+  // upgradeMHVAccount,
```
I think you can just remove this line, right?

U-DON commented on src/applications/validate-mhv-account/containers/ErrorMessage.jsx:
```diff
@@ -0,0 +1,57 @@
+import PropTypes from 'prop-types';
+import { connect } from 'react-redux';
+import {
+  createAndUpgradeMHVAccount,
```
It doesn't seem like this actually ends up getting used in this component.

U-DON commented on src/applications/validate-mhv-account/components/errors/UpgradeAccountFailed.jsx:
```diff
@@ -0,0 +1,103 @@
+import PropTypes from 'prop-types';
+import MessageTemplate from '../MessageTemplate';
+
+const UpgradeAccountFailed = ({ upgradeAccount }) => {
```
Should this be more consistent with `CreateMHVAccount` by making it a container that maps `upgradeMHVAccount` to its own prop?

U-DON commented on src/applications/validate-mhv-account/containers/VerifyIdentity.jsx:
```diff
@@ -0,0 +1,28 @@
+import MessageTemplate from './../components/MessageTemplate';
+import { verify } from '../../../platform/user/authentication/utilities';
+
+const MultipleMHVIds = () => {
```
I think name may have been accidentally copy-pasted?

U-DON commented on src/applications/validate-mhv-account/containers/ValidateMHVAccount.jsx:
```diff
@@ -0,0 +1,111 @@
+import LoadingIndicator from '@department-of-veterans-affairs/formation-react/LoadingIndicator';
+
+import {
+  createAndUpgradeMHVAccount,
```
It doesn't seem like this ends up getting used here.

U-DON commented on src/applications/validate-mhv-account/components/errors/NeedsVAPatient.jsx:
```diff
@@ -0,0 +1,72 @@
+import React from 'react';
+import MessageTemplate from '../MessageTemplate';
+
+const NeedaSSNResolution = () => {
```
I think this is supposed to be `NeedsVAPatient`.

narin commented:
@U-DON Thanks for the review.  I've addressed your suggestions and you can find them in this PR: https://github.com/department-of-veterans-affairs/vets-website/pull/9996.  Sorry for the multiple PRs.  This one should be the final one.
